src/fundus/scraping/article.py

```diff
@@ -16,37 +14,76 @@
 logger = create_logger(__name__)
 
 
-@dataclass(frozen=True)
+class AttributeView:
+    def __init__(self, key: str, extraction: Mapping[str, Any]):
+        self.ref = extraction
+        self.key = key
+
+    def __get__(self, instance: object, owner: type):
+        return self.ref[self.key]
+
+    def __set__(self, obj, value):
+        # For now, this is read-only
+        raise AttributeError("attribute is read only")
+
+
 class Article:
-    html: HTML
-    exception: Optional[Exception] = None
-
-    # supported (validated) attributes as defined in the guidelines
-    title: Optional[str] = None
-    authors: List[str] = field(default_factory=list)
-    body: Optional[ArticleBody] = None
-    publishing_date: Optional[datetime] = None
-    topics: List[str] = field(default_factory=list)
-    free_access: bool = True
-
-    @classmethod
-    def from_extracted(cls, html: HTML, extracted: Dict[str, Any], exception: Optional[Exception] = None) -> "Article":
-        validated_attributes: Set[str] = {article_field.name for article_field in fields(cls)}
-
-        extracted_unvalidated: Iterator[Tuple[str, Any]]
-        extracted_validated: Iterator[Tuple[str, Any]]
-        extracted_unvalidated, extracted_validated = more_itertools.partition(
-            lambda attribute_and_value: attribute_and_value[0] in validated_attributes, extracted.items()
-        )
+    __extraction__: Mapping[str, Any] = {}
+
+    def __init__(self, *, html: HTML, exception: Optional[Exception] = None, **extraction: Any) -> None:
+        self.html = html
+        self.exception = exception
+        self.__extraction__ = extraction
 
-        article: Article = cls(html, exception, **dict(extracted_validated))
-        unvalidated_attributes: Set[str] = set()
-        for attribute, value in extracted_unvalidated:
-            object.__setattr__(article, attribute, value)  # Sets attributes on a frozen dataclass
-            unvalidated_attributes.add(attribute)
-        object.__setattr__(article, "_unvalidated_attributes", unvalidated_attributes)
+        # create descriptors for attributes that aren't pre-defined as properties.
+        for attribute in extraction.keys():
+            if not hasattr(self, attribute):
+                setattr(self, attribute, AttributeView(attribute, self.__extraction__))
+
```

Comment:
If we're already making properties for all the important properties of the article, I would also consider adding one for the responded url.

Reply:
Hmm, I think then we also have to do one for requested URL and that's getting a bit overwhelming. I totally get your point here and I think you're totally right, dunno yet if that's the best solution.

Reply:
Ok, then we can just keep it in mind for future updates